homeassistant/components/climate/__init__.py

```diff
@@ -288,6 +300,94 @@ class ClimateEntity(Entity, cached_properties=CACHED_PROPERTIES_WITH_ATTR_):
+
+    def __getattribute__(self, __name: str) -> Any:
+        """Get attribute.
+
+        Modify return of `supported_features` to
+        include `_mod_supported_features` if attribute is set.
+        """
+        if __name != "supported_features":
+            return super().__getattribute__(__name)
+
+        # Implements check for deprecated use of ClimateEntityFeature as int
+        _supported_features = super().__getattribute__(__name)
+        if type(_supported_features) is int:  # noqa: E721
+            new_features = ClimateEntityFeature(_supported_features)
+            self._report_deprecated_supported_features_values(new_features)
+        return _supported_features | super().__getattribute__(
+            "_ClimateEntity__mod_supported_features"
+        )
+
+    @callback
+    def add_to_platform_start(
+        self,
+        hass: HomeAssistant,
+        platform: EntityPlatform,
+        parallel_updates: asyncio.Semaphore | None,
+    ) -> None:
+        """Start adding an entity to a platform."""
+        super().add_to_platform_start(hass, platform, parallel_updates)
+
+        def _report_turn_on_off(feature: str, method: str) -> None:
+            """Log warning not implemented turn on/off feature."""
+            report_issue = self._suggest_report_issue()
+            if feature.startswith("TURN"):
+                message = (
+                    "Entity %s (%s) does not set ClimateEntityFeature.%s"
+                    " but implements the %s method. Please %s"
+                )
+            else:
+                message = (
+                    "Entity %s (%s) implements HVACMode.%s and therefore implicitly"
+                    " implements the %s method without setting the proper"
+                    " ClimateEntityFeature. Please %s"
+                )
+            _LOGGER.warning(
+                message,
+                self.entity_id,
+                type(self),
+                feature,
+                feature.lower(),
+                report_issue,
+            )
+
+        # Adds ClimateEntityFeature.TURN_OFF/TURN_ON depending on service calls implemented
+        # This should be removed in 2025.1.
+        if not self.supported_features & ClimateEntityFeature.TURN_OFF:
+            if HVACMode.OFF in self.hvac_modes and (
+                type(self).async_turn_off is not ClimateEntity.async_turn_off
+                or type(self).turn_off is not ClimateEntity.turn_off
+            ):
+                _report_turn_on_off("TURN_OFF", "turn_off")
+                self.__mod_supported_features |= (  # pylint: disable=unused-private-member
+                    ClimateEntityFeature.TURN_OFF
+                )
+            elif HVACMode.OFF in self.hvac_modes:
+                _report_turn_on_off("off", "turn_off")
+                self.__mod_supported_features |= (  # pylint: disable=unused-private-member
+                    ClimateEntityFeature.TURN_OFF
+                )
```

Comment:
The logic does not appear to be correct, we should automatically set `ClimateEntityFeature.TURN_OFF` if *either* (inclusive) of the following is true:
- The derived class overrides the default implementation of climate.turn_on or climate.turn_off
- The required HVAC modes are supported